File: aerox.Runtime.Widgets/WidgetWindowSurface.cs
using aerox.Runtime.Graphics;
using aerox.Runtime.Math;
using aerox.Runtime.Widgets.Events;
using aerox.Runtime.Windows;
using TerraFX.Interop.Vulkan;

namespace aerox.Runtime.Widgets;

/// <summary>
///     A surface that displays widgets on a window
/// </summary>
public class WidgetWindowSurface : WidgetSurface
{
    private readonly WindowRenderer _renderer;
    public bool Minimized;
    public Vector2<int> Size;

    public WidgetWindowSurface(WindowRenderer renderer)
    {
        _renderer = renderer;
        Size = new Vector2<int>((int)Window.PixelSize.width, (int)Window.PixelSize.height);
    }

    public Window Window => _renderer.GetWindow();

    public void CopyToSwapchain(Frame frame, VkImage swapchainImage, VkExtent2D extent)
    {
        var drawImage = GetDrawImage();
        SGraphicsModule.CopyImageToImage(frame.GetCommandBuffer(), drawImage.Image, swapchainImage,
            drawImage.Extent, _renderer.GetSwapchainExtent());
    }

    public override void Init()
    {
        base.Init();
        Window.OnResized += OnWindowResize;
        Window.OnMouseButton += OnMouseButton;
        Window.OnMouseMoved += OnMouseMove;
        Window.OnScrolled += OnScroll;
        _renderer.OnDrawSecondary += Draw;
        _renderer.OnCopyToSwapchain += CopyToSwapchain;
    }

    protected void OnWindowResize(Window.ResizeEvent e)
    {
        Size = new Vector2<int>((int)e.Width, (int)e.Height);
        Minimized = Size.X == 0 || Size.Y == 0;
        if (!Minimized) ReceiveResize(new ResizeEvent(this, Size.Clone()));
    }

    protected void OnMouseButton(Window.MouseButtonEvent e)
    {
        switch (e.State)
        {
            case EKeyState.Released:
                ReceiveCursorUp(new CursorUpEvent(this, e.Position.Cast<float>()));
                break;
            case EKeyState.Pressed:
                ReceiveCursorDown(new CursorDownEvent(this, e.Position.Cast<float>()));
                break;
            case EKeyState.Repeat:
                break;
            default:
                throw new ArgumentOutOfRangeException();
        }
    }

    protected void OnMouseMove(Window.MouseMoveEvent e)
    {
        ReceiveCursorMove(new CursorMoveEvent(this, e.Position.Cast<float>()));
    }

    protected void OnScroll(Window.ScrollEvent e)
    {
        ReceiveScroll(new ScrollEvent(this, e.Position.Cast<float>(), e.Delta.Cast<float>()));
    }


    protected override void OnDispose(bool isManual)
    {
        base.OnDispose(isManual);
        Window.OnResized -= OnWindowResize;
        Window.OnMouseButton -= OnMouseButton;
        Window.OnMouseMoved -= OnMouseMove;
        Window.OnScrolled -= OnScroll;

        _renderer.OnDrawSecondary -= Draw;
        _renderer.OnCopyToSwapchain -= CopyToSwapchain;
    }

    public override void Draw(Frame frame)
    {
        if (!Minimized) base.Draw(frame);
    }

    public override Vector2<float> GetCursorPosition()
    {
        return Window.GetMousePosition().Cast<float>();
    }

    public override Vector2<int> GetDrawSize()
    {
        return Size;
    }
}

File: modules/widgets/lib/rin/widgets/WidgetWindowSurface.cpp
#include "rin/widgets/WidgetWindowSurface.hpp"

#include "rin/widgets/event/CursorDownEvent.hpp"
#include "rin/widgets/event/CursorMoveEvent.hpp"
#include "rin/widgets/event/CursorUpEvent.hpp"
#include "rin/widgets/event/ResizeEvent.hpp"
WidgetWindowSurface::WidgetWindowSurface(WindowRenderer* window)
    {
        _windowRenderer = window;
    }

    Vec2<int> WidgetWindowSurface::GetDrawSize() const
    {
        return _windowRenderer->GetWindow()->GetSize();
    }

    Vec2<float> WidgetWindowSurface::GetCursorPosition() const
    {
        return _windowRenderer->GetWindow()->GetCursorPosition();
    }

    void WidgetWindowSurface::Init()
    {
        WidgetSurface::Init();
        auto thisShared = this->GetSharedDynamic<WidgetWindowSurface>();
        auto window = _windowRenderer->GetWindow();
        _drawHandle = _windowRenderer->onDraw->Add<WidgetWindowSurface>(thisShared,&WidgetWindowSurface::Draw);
        _copyHandle = _windowRenderer->onCopy->Add<WidgetWindowSurface>(thisShared,&WidgetWindowSurface::CopyToSwapchain);
        _resizeHandle = window->onResize->Add<WidgetWindowSurface>(thisShared,&WidgetWindowSurface::OnWindowResize);
        _cursorMoveHandle = window->onCursorMoved->Add<WidgetWindowSurface>(thisShared,&WidgetWindowSurface::OnWindowCursorMove);
        _cursorButtonHandle = window->onCursorButton->Add<WidgetWindowSurface>(thisShared,&WidgetWindowSurface::OnWindowCursorButton);
    }

    void WidgetWindowSurface::HandleDrawSkipped(Frame* frame)
    {
        if(auto image = GetDrawImage())
        {
            image->Barrier(frame->GetCommandBuffer(),vk::ImageLayout::eUndefined,vk::ImageLayout::eTransferSrcOptimal);
        }

        if(auto image = GetCopyImage())
        {
            image->Barrier(frame->GetCommandBuffer(),vk::ImageLayout::eUndefined,vk::ImageLayout::eTransferSrcOptimal);
        }
    }

    void WidgetWindowSurface::HandleBeforeDraw(Frame* frame)
    {
        if(auto image = GetDrawImage())
        {
            image->Barrier(frame->GetCommandBuffer(),vk::ImageLayout::eUndefined,vk::ImageLayout::eColorAttachmentOptimal);
        }

        if(auto image = GetCopyImage())
        {
            image->Barrier(frame->GetCommandBuffer(),vk::ImageLayout::eUndefined,vk::ImageLayout::eColorAttachmentOptimal);
        }
    }

    void WidgetWindowSurface::HandleAfterDraw(Frame* frame)
    {
        if(auto image = GetDrawImage())
        {
            image->Barrier(frame->GetCommandBuffer(),vk::ImageLayout::eColorAttachmentOptimal,vk::ImageLayout::eTransferSrcOptimal);
        }
    }

    void WidgetWindowSurface::CopyToSwapchain(Frame* frame, const vk::Image& dest, const vk::Extent2D& destExtent)
    {
        if(auto drawImage = GetDrawImage())
        {
            drawImage->CopyTo(frame->GetCommandBuffer(),dest,{destExtent.width,destExtent.height,1});
        }
    }

    void WidgetWindowSurface::OnWindowResize(Window * window,const Vec2<int>& size)
    {
        NotifyResize(newShared<ResizeEvent>(this->GetSharedDynamic<WidgetWindowSurface>(),size.Cast<float>()));
    }

    void WidgetWindowSurface::OnWindowCursorMove(Window* window, const Vec2<float>& position)
    {
        NotifyCursorMove(newShared<CursorMoveEvent>(this->GetSharedDynamic<WidgetWindowSurface>(),position));
    }

    void WidgetWindowSurface::OnWindowCursorButton(Window* window,  CursorButton button, InputState state)
    {
        if(state == InputState::Pressed)
        {
            NotifyCursorDown(newShared<CursorDownEvent>(this->GetSharedDynamic<WidgetWindowSurface>(),button,GetCursorPosition()));
        }
        else if(state == InputState::Released)
        {
            NotifyCursorUp(newShared<CursorUpEvent>(this->GetSharedDynamic<WidgetWindowSurface>(),button,GetCursorPosition()));
        }
    }

    void WidgetWindowSurface::OnDispose(bool manual)
    {
        WidgetSurface::OnDispose(manual);
        _drawHandle.UnBind();
        _copyHandle.UnBind();
        _resizeHandle.UnBind();
        _cursorMoveHandle.UnBind();
        _cursorButtonHandle.UnBind();
    }
